File: packages/SystemUI/src/com/android/systemui/communal/dagger/CommunalModule.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.communal.dagger;

import android.content.ComponentName;
import android.content.Context;
import android.content.res.Resources;
import android.text.TextUtils;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;

import com.android.systemui.R;
import com.android.systemui.communal.CommunalSource;
import com.android.systemui.communal.PackageObserver;
import com.android.systemui.communal.conditions.CommunalSettingCondition;
import com.android.systemui.dagger.qualifiers.Main;
import com.android.systemui.idle.AmbientLightModeMonitor;
import com.android.systemui.idle.LightSensorEventsDebounceAlgorithm;
import com.android.systemui.idle.dagger.IdleViewComponent;
import com.android.systemui.util.condition.Condition;
import com.android.systemui.util.condition.Monitor;
import com.android.systemui.util.condition.dagger.MonitorComponent;

import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import javax.inject.Named;
import javax.inject.Provider;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.multibindings.ElementsIntoSet;
import dagger.multibindings.IntoMap;
import dagger.multibindings.StringKey;

/**
 * Dagger Module providing Communal-related functionality.
 */
@Module(subcomponents = {
        CommunalViewComponent.class,
        IdleViewComponent.class,
})
public interface CommunalModule {
    String IDLE_VIEW = "idle_view";
    String COMMUNAL_CONDITIONS = "communal_conditions";

    /** */
    @Provides
    @Named(IDLE_VIEW)
    static View provideIdleView(Context context) {
        FrameLayout view = new FrameLayout(context);
        return view;
    }

    /** */
    @Provides
    static Optional<CommunalSource.Observer> provideCommunalSourcePackageObserver(
            Context context, @Main Resources resources) {
        final String componentName = resources.getString(R.string.config_communalSourceComponent);

        if (TextUtils.isEmpty(componentName)) {
            return Optional.empty();
        }

        return Optional.of(new PackageObserver(context,
                ComponentName.unflattenFromString(componentName).getPackageName()));
    }

    /**
     * Provides LightSensorEventsDebounceAlgorithm as an instance to DebounceAlgorithm interface.
     * @param algorithm the instance of algorithm that is bound to the interface.
     * @return the interface that is bound to.
     */
    @Binds
    AmbientLightModeMonitor.DebounceAlgorithm ambientLightDebounceAlgorithm(
            LightSensorEventsDebounceAlgorithm algorithm);

    /**
     * Provides a set of conditions that need to be fulfilled in order for Communal Mode to display.
     */
    @Provides
    @ElementsIntoSet
    @Named(COMMUNAL_CONDITIONS)
    static Set<Condition> provideCommunalConditions(
            CommunalSettingCondition communalSettingCondition) {
        return new HashSet<>(Collections.singletonList(communalSettingCondition));
    }

    /**
     * TODO(b/205638389): Remove when there is a base implementation of
     * {@link CommunalSource.Connector}. Currently a place holder to allow a map to be present.
     */
    @Provides
    @IntoMap
    @Nullable
    @StringKey("empty")
    static CommunalSource.Connector provideEmptyCommunalSourceConnector() {
        return null;
    }

    /** */
    @Provides
    static Optional<CommunalSource.Connector> provideCommunalSourceConnector(
            @Main Resources resources,
            Map<Class<?>, Provider<CommunalSource.Connector>> connectorCreators) {
        final String className = resources.getString(R.string.config_communalSourceConnector);

        if (TextUtils.isEmpty(className)) {
            return Optional.empty();
        }

        try {
            Class<?> clazz = Class.forName(className);
            Provider<CommunalSource.Connector> provider = connectorCreators.get(clazz);
            return provider != null ? Optional.of(provider.get()) : Optional.empty();
        } catch (ClassNotFoundException e) {
            return Optional.empty();
        }
    }

    /** */
    @Provides
    @Named(COMMUNAL_CONDITIONS)
    static Monitor provideCommunalSourceMonitor(
            @Named(COMMUNAL_CONDITIONS) Set<Condition> communalConditions,
            MonitorComponent.Factory factory) {
        final MonitorComponent component = factory.create(communalConditions, new HashSet<>());
        return component.getMonitor();
    }
}

File: packages/SystemUI/src/com/android/systemui/communal/dagger/CommunalModule.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.communal.dagger

import android.content.Context
import android.content.res.Resources
import com.android.systemui.CoreStartable
import com.android.systemui.communal.data.backup.CommunalBackupUtils
import com.android.systemui.communal.data.db.CommunalDatabaseModule
import com.android.systemui.communal.data.repository.CommunalMediaRepositoryModule
import com.android.systemui.communal.data.repository.CommunalPrefsRepositoryModule
import com.android.systemui.communal.data.repository.CommunalRepositoryModule
import com.android.systemui.communal.data.repository.CommunalSettingsRepositoryModule
import com.android.systemui.communal.data.repository.CommunalSmartspaceRepositoryModule
import com.android.systemui.communal.data.repository.CommunalTutorialRepositoryModule
import com.android.systemui.communal.data.repository.CommunalWidgetRepositoryModule
import com.android.systemui.communal.domain.interactor.CommunalSceneTransitionInteractor
import com.android.systemui.communal.shared.log.CommunalMetricsLogger
import com.android.systemui.communal.shared.log.CommunalStatsLogProxyImpl
import com.android.systemui.communal.shared.model.CommunalScenes
import com.android.systemui.communal.util.CommunalColors
import com.android.systemui.communal.util.CommunalColorsImpl
import com.android.systemui.communal.widgets.CommunalWidgetModule
import com.android.systemui.communal.widgets.EditWidgetsActivityStarter
import com.android.systemui.communal.widgets.EditWidgetsActivityStarterImpl
import com.android.systemui.dagger.SysUISingleton
import com.android.systemui.dagger.qualifiers.Application
import com.android.systemui.dagger.qualifiers.Main
import com.android.systemui.res.R
import com.android.systemui.scene.shared.model.SceneContainerConfig
import com.android.systemui.scene.shared.model.SceneDataSource
import com.android.systemui.scene.shared.model.SceneDataSourceDelegator
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.multibindings.ClassKey
import dagger.multibindings.IntoMap
import javax.inject.Named
import kotlinx.coroutines.CoroutineScope

@Module(
    includes =
        [
            CommunalRepositoryModule::class,
            CommunalMediaRepositoryModule::class,
            CommunalTutorialRepositoryModule::class,
            CommunalWidgetRepositoryModule::class,
            CommunalDatabaseModule::class,
            CommunalWidgetModule::class,
            CommunalPrefsRepositoryModule::class,
            CommunalSettingsRepositoryModule::class,
            CommunalSmartspaceRepositoryModule::class,
            CommunalStartableModule::class,
        ]
)
interface CommunalModule {
    @Binds
    fun bindEditWidgetsActivityStarter(
        starter: EditWidgetsActivityStarterImpl
    ): EditWidgetsActivityStarter

    @Binds
    @Communal
    fun bindCommunalSceneDataSource(@Communal delegator: SceneDataSourceDelegator): SceneDataSource

    @Binds fun bindCommunalColors(impl: CommunalColorsImpl): CommunalColors

    @Binds
    fun bindCommunalStatsLogProxy(
        impl: CommunalStatsLogProxyImpl
    ): CommunalMetricsLogger.StatsLogProxy

    @Binds
    @IntoMap
    @ClassKey(CommunalSceneTransitionInteractor::class)
    abstract fun bindCommunalSceneTransitionInteractor(
        impl: CommunalSceneTransitionInteractor
    ): CoreStartable

    companion object {
        const val LOGGABLE_PREFIXES = "loggable_prefixes"
        const val LAUNCHER_PACKAGE = "launcher_package"

        @Provides
        @Communal
        @SysUISingleton
        fun providesCommunalSceneDataSourceDelegator(
            @Application applicationScope: CoroutineScope
        ): SceneDataSourceDelegator {
            val config =
                SceneContainerConfig(
                    sceneKeys = listOf(CommunalScenes.Blank, CommunalScenes.Communal),
                    initialSceneKey = CommunalScenes.Blank,
                    navigationDistances =
                        mapOf(
                            CommunalScenes.Blank to 0,
                            CommunalScenes.Communal to 1,
                        ),
                )
            return SceneDataSourceDelegator(applicationScope, config)
        }

        @Provides
        @SysUISingleton
        fun providesCommunalBackupUtils(
            @Application context: Context,
        ): CommunalBackupUtils {
            return CommunalBackupUtils(context)
        }

        /** The prefixes of widgets packages names that are considered loggable. */
        @Provides
        @Named(LOGGABLE_PREFIXES)
        fun provideLoggablePrefixes(@Application context: Context): List<String> {
            return context.resources
                .getStringArray(com.android.internal.R.array.config_loggable_dream_prefixes)
                .toList()
        }

        /** The package name of the launcher */
        @Provides
        @Named(LAUNCHER_PACKAGE)
        fun provideLauncherPackage(@Main resources: Resources): String {
            return resources.getString(R.string.launcher_overlayable_package)
        }
    }
}
